Task/backend36 image upload

WanderningMaster opened the pull request:
I added test controller to make sure that images are saved in cloud

WanderningMaster committed:
backend-36: + cloudinary package and credentials to config

WanderningMaster committed:
backedn-36: + cloudinary container module

WanderningMaster committed:
backend-36: + cloudinary adapter

WanderningMaster committed:
backend-36: * binded cloudinary adapter to container module

WanderningMaster committed:
backend-36: + cloudinary adapter to user repository

WanderningMaster committed:
backend-36: - unnecessary cloudinary method for getting image from store, + test controller

WanderningMaster committed:
backend-36: * Resolve merge conflicts

WanderningMaster commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement image uploading logic](https://trello.com/c/sIjELRJT/36-implement-image-uploading-logic)

VitaliiBedletskyi commented on backend/src/configuration/config.ts:
```diff
@@ -21,9 +21,16 @@ interface ApiConfig {
 export interface ConfigInterface {
   APP: AppConfig;
   DATABASE: DatabaseConfig;
+  CLOUD: CloudConfig;
```
Probably, more descriptive name should be used like: `IMAGE_CLOUD_STORAGE`

VitaliiBedletskyi commented on backend/src/core/user/port/user-repository.ts:
```diff
@@ -1,8 +1,11 @@
 
   createUser(userRequestDto: UserSignUpRequestDto): Promise<UserSignUpResponseDto>;
+
+  uploadAvatar(base64Str: string): Promise<UploadApiResponse>;
```
You should not mix a repository for working with a database and a service (adapter) for working with the cloud for storing image files. It should be separate files.

VitaliiBedletskyi replied:
In your case should directly inject in UserService the CloudinaryAdapter that you already created.

VitaliiBedletskyi commented on backend/src/core/common/port/cloudinary.ts:
```diff
@@ -0,0 +1,6 @@
+import { UploadApiResponse } from 'cloudinary';
+import { CloudinaryUploadDto } from '~/shared/types/types';
+
+export interface CloudinaryPort {
```
You shouldn't use a cloud name, it's better to abstract away the names of the services that we use in our business logic. It should be something like `ImageStoragePort`

VitaliiBedletskyi commented on backend/src/core/common/port/cloudinary.ts:
```diff
@@ -0,0 +1,6 @@
+import { UploadApiResponse } from 'cloudinary';
```
Better to change to custom DTO with only needed fields. Port shouldn't know anything about the service which we use.

VitaliiBedletskyi commented on backend/src/core/common/port/cloudinary.ts:
```diff
@@ -0,0 +1,6 @@
+import { CloudinaryUploadDto } from '~/shared/types/types';
+
+export interface CloudinaryPort {
+  upload(uploadDto: CloudinaryUploadDto): Promise<UploadApiResponse>;
```
The same thing as above, rename...

VitaliiBedletskyi commented on backend/src/secondary-adapters/cloudinary/cloudinary-container-module.ts:
```diff
@@ -0,0 +1,25 @@
+import { CloudinaryPort } from '~/core/common/port/cloudinary';
+import { CloudinaryAdapter } from './cloud/clodinary-adapter';
+
+const cloudinaryContainerModule = new AsyncContainerModule(async (bind: interfaces.Bind) => {
```
Why do you need to use this AsyncContainerModule instead of ContainerModule. I don't see any asynchronous operations... Did I miss something?

WanderningMaster replied:
I guess we need since i moved whole logic concerning to cloudinary to secondary adapters
![image](https://user-images.githubusercontent.com/56759981/184027493-48e20d96-4be5-43fd-b85d-931383db4817.png)

VitaliiBedletskyi commented on backend/src/shared/enums/cloudinary/cloudinary.ts:
```diff
@@ -0,0 +1 @@
+export { CloudinaryPresetType } from 'shared/build';
```
Get rid of `cloudinary` name in filenames and types

VitaliiBedletskyi commented on shared/src/common/types/cloudinary/cloudinary-upload-dto.type.ts:
```diff
@@ -0,0 +1,8 @@
+import { CloudinaryPresetType } from '~/common/enums/cloudinary/cloudinary-preset.enum';
+
+type CloudinaryUploadDto = {
```
The same problem with the name...

VitaliiBedletskyi commented:
FYI: please resolve merge conflicts with the `development` branch before updating PR.

WanderningMaster committed:
backend-36: + dto types for upload route, + dto for user avatar upload response

WanderningMaster committed:
backend-36: * resolve merge conflicts

WanderningMaster committed:
backend-36: * moved cloudinary service to user service

WanderningMaster committed:
backend-36: * cloud naming to image store to encapsulate realisation from business logic,
+ npm script for docker to up only services like db, message broker etc. so we can debug app

WanderningMaster committed:
backend-36: * fixed import in cloud adapter

WanderningMaster committed:
backend-36: * resolved types naming,
* moved whole logic related with cloudinary to one container,
+ jsdocs for /upload route

VitaliiBedletskyi commented on backend/src/secondary-adapters/cloudinary/cloudinary-container-module.ts:
```diff
@@ -0,0 +1,25 @@
+import { ImageStorePort } from '~/core/common/port/image-store';
+import { ImageStoreAdapter } from './cloud/clodinary-adapter';
+
+const cloudinaryContainerModule = new AsyncContainerModule(async (bind: interfaces.Bind) => {
```
Probably you need to use ContainerModule instead of AsyncContainerModule

WanderningMaster committed:
backend-36: * changed from async to pure module container

WanderningMaster committed:
backend-36: * Resolve merge conflicts

WanderningMaster committed:
Merge branch 'development' into task/backend36-image-upload
